dash_auth/oidc_auth.py

```diff
@@ -0,0 +1,298 @@
+        :param kwargs: Keyword arguments passed to OAuth.register.
+            See https://docs.authlib.org/en/latest/client/flask.html for
+            additional details.
+            Typical keyword arguments for OIDC include:
+            * client_id
+            * client_secret
+            * server_metadata_url
+            * token_endpoint_auth_method
+            * client_kwargs (defaults to {"scope": "openid email"})
+        """
+        client_kwargs = kwargs.pop("client_kwargs", {})
+        client_kwargs.setdefault("scope", "openid email")
+        self.oauth.register(
+            idp_name, client_kwargs=client_kwargs, **kwargs
+        )
+
+    def get_oauth_client(self, idp: str):
+        """Get the OAuth client."""
+        if not idp:
+            raise ValueError("`idp` should be set")
+
+        client: Union[FlaskOAuth1App, FlaskOAuth2App] = (
+            self.oauth.create_client(idp)
+        )
+        return client
+
+    def get_oauth_kwargs(self, idp: str = None):
+        """Get the OAuth kwargs."""
+        if not idp:
+            raise ValueError("`idp` should be set")
+
+        kwargs: dict = (
+            self.oauth._registry[idp][1]
+        )
+        return kwargs
+
+    def login_request(self, idp: str = None):
+        """Login the user."""
+        if idp is None:
+            if len(self.oauth._clients) == 1:
+                idp = next(iter(self.oauth._clients))
+            elif self.idp_selection_route:
+                return redirect(self.idp_selection_route)
+            else:
+                return (
+                    "Several OAuth providers are registered. "
+                    "Please choose one.",
+                    400,
+                )
+
+        kwargs = {"_external": True}
+        if self.force_https_callback:
+            kwargs["_scheme"] = "https"
+        redirect_uri = url_for("oidc_callback", idp=idp, **kwargs)
+        if request.headers.get("X-Forwarded-Host"):
+            host = request.headers.get("X-Forwarded-Host")
+            redirect_uri = redirect_uri.replace(request.host, host)
+
+        oauth_client = self.get_oauth_client(idp)
+        oauth_kwargs = self.get_oauth_kwargs(idp)
+        return oauth_client.authorize_redirect(
+            redirect_uri,
+            **oauth_kwargs.get("authorize_redirect_kwargs", {}),
+        )
+
+    def logout(self):  # pylint: disable=C0116
+        """Logout the user."""
+        session.clear()
+        base_url = self.app.config.get("url_base_pathname") or "/"
+        page = self.logout_page or f"""
+        <div style="display: flex; flex-direction: column;
+        gap: 0.75rem; padding: 3rem 5rem;">
+            <div>Logged out successfully</div>
+            <div><a href="{base_url}">Go back</a></div>
+        </div>
+        """
+        return page, 200
+
+    def callback(self, idp: str = None):  # pylint: disable=C0116
+        """Do the OIDC dance."""
+        if idp is None:
```

Comment:
Can idp be None at all? the callback URL is formed in the login_request which guards against it I think?

Reply:
Not in the normal flow of login_request -> idp -> callback but this could theoretically be used in other places.

Fo our usecase for instance since we already gave clients the redirect_uri `/callback` we could create a flask route `/callback` that calls OIDCAuth.callback. It would have None as the idp but it would be fine as there is only 1 IDP registered.

Reply:
I feel like exiting when idp is None (or actually when idp is not a valid key in the registry) is the right way to handle it here.

If downstream applications need to define redirects for legacy reasons, that's fine: they can define the idp in their redirect